src/shared/tpm2-util.c

```diff
@@ -7031,6 +7170,74 @@ int tpm2_load_public_key_file(const char *path, TPM2B_PUBLIC *ret) {
         *ret = device_key_public;
         return 0;
 }
+
+int tpm2_hmac_key_from_pin(Tpm2Context *c, const Tpm2Handle *session, const TPM2B_AUTH *pin, Tpm2Handle **ret) {
+        int r;
+
+        assert(c);
+        assert(pin);
+        assert(ret);
+
+        /* Load the PIN (which we have stored in the "auth" TPM2B_AUTH) into the TPM as an HMAC key so that
+         * we can use it in a TPM2_PolicySigned() to write to the nvindex. For that we'll prep a pair of
+         * TPM2B_PUBLIC and TPM2B_SENSITIVE that defines an HMAC-SHA256 keyed hash function, and initialize
+         * it based on on the provided PIN data. */
+
+        TPM2B_PUBLIC auth_hmac_public = {
+                .publicArea = {
+                        .type = TPM2_ALG_KEYEDHASH,
+                        .nameAlg = TPM2_ALG_SHA256,
+                        .objectAttributes = TPMA_OBJECT_SIGN_ENCRYPT | TPMA_OBJECT_USERWITHAUTH,
+                        .parameters.keyedHashDetail.scheme = {
```

Comment:
You never set the authValue, which means anyone could use this key is that intentional?

Also, without fixedParent set, someone could use TPM2_Duplicate to get access to the internal key material.

Now that I understand the lifecycle of this key and it's usages, I think you want to drop userwithauth and set fixedParent.

Reply:
hmm, this is the HMAC key we load into the NULL hierarchy only for a short period of time. We are not going to do much with it, except for using it in TPM2_PolicySigned(). 

I guess we could set authValue to something random so that noone can use the key for anything if they happen to get ahold of the handle while we loaded it into the TPM.

does fixedParent work in the NULL hierarchy?

Reply:
> hmm, this is the HMAC key we load into the NULL hierarchy only for a short period of time. We are not going to do much with it, except for using it in TPM2_PolicySigned().
> 
> I guess we could set authValue to something random so that noone can use the key for anything if they happen to get ahold of the handle while we loaded it into the TPM.
> 
> does fixedParent work in the NULL hierarchy?

Oh I missed it's loaded in NULL, that's why unique doesn't matter. 

If you just drop userwithauth it should be fine enough. It was just a bonus to randomize the auth value.

You could load this under the SRK as a proper object using tpm2 import, but that's a lot of work as you need to perform object wrapping using the SRK public. We need to create a utility library with this function, in it's the python code base.